### PR TITLE
Mstp yang

### PR DESCRIPTION
**Pull request by wajahatrazi**

### CODE REVIEW

File: src/sonic-yang-models/yang-models/sonic-stp.yang
Comment (divyachandralekha) [inline src/sonic-yang-models/yang-models/sonic-stp.yang:1]: Please include pvst fileds also.

Comment (wajahatrazi) [reply src/sonic-yang-models/yang-models/sonic-stp.yang:1]: Hi Divya, I have added the changes for both pvst & mstp.

File: src/sonic-yang-models/yang-models/sonic-stp.yang [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+                description "STP mode (PVST or MST)";
+            }
+            leaf rootguard_timeout {
+                type uint16 {
+                    range "5..600";
+                }
+                default "30";
+                description "Root guard timeout in seconds";
+            }
+            leaf forward_delay {
```
Comment (divyachandralekha) [inline src/sonic-yang-models/yang-models/sonic-stp.yang]: grouping vlanModeAttr {
        leaf forward_delay {
            type uint8 {
                range "4..30" {
                    error-message "Invalid Forwarding Delay value.";
                }
            }
            units seconds;
            default 15;
            description
                "The delay used by STP bridges to transition root and
                designated ports to forwarding";
        }

        leaf hello_time {
            type uint8 {
                range "1..10" {
                    error-message "Invalid Hello Time value.";
                }
            }
            units seconds;
            default 2;
            description
                "The interval between periodic transmissions of
                configuration messages by designated ports";
        }

        leaf max_age {
            type uint8 {
                range "6..40" {
                    error-message "Invalid Maximum Age Time value.";
                }
            }
            units seconds;
            default 20;
            description
                "The maximum age of the information transmitted by the
                bridge when it is the root bridge";
        }

        leaf priority {
            type uint16 {
                range "0..61440" {
                    error-message "Invalid Bridge Priority value.";
                }
            }
            sonic-ext:custom-validation ValidateBridgePriority;
            default 32768;
            description
                "The manageable component of the Bridge Identifier";
        }
    }

Add this , make add as below on this block
  uses vlanModeAttr;

File: src/sonic-yang-models/yang-models/sonic-stp.yang [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+                description "Maximum age time in seconds";
+            }
+            leaf priority {
+                type uint16 {
+                    range "0..61440";
+                }
+                default "32768";
+                description "Bridge priority";
+            }
+
```
Comment (divyachandralekha) [inline src/sonic-yang-models/yang-models/sonic-stp.yang:177]: Close the bracket here. MST block should be as below
      container STP_MST {
            list STP_MST_LIST {
                max-elements 1;
                key "keyleaf";
                sonic-ext:dependent-on "STP_LIST";

                leaf keyleaf {
                    type enumeration {
                        enum GLOBAL;
                    }
                    description
                        "Key node identifier. It's value is always GLOBAL";
                }

                leaf name {
                    must "../../../STP/STP_LIST[keyleaf='GLOBAL']/mode='mst'" {
                        error-message "Configuration allowed in MST mode";
                        error-app-tag stp-invalid;
                    }
                    type string;
                    description
                        "MST Region name";
                }

                leaf revision {
                    must "../../../STP/STP_LIST[keyleaf='GLOBAL']/mode='mst'" {
                        error-message "Configuration allowed in MST mode";
                        error-app-tag stp-invalid;
                    }
                    type uint32;
                    description
                        "MST Revision number";
                }

                leaf max_hops {
                    must "../../../STP/STP_LIST[keyleaf='GLOBAL']/mode='mst'" {
                        error-message "Configuration allowed in MST mode";
                        error-app-tag stp-invalid;
                    }
                    type uint8;
                    description
                        "MST Max hops";
                }

                leaf hello_time {
                    must "../../../STP/STP_LIST[keyleaf='GLOBAL']/mode='mst'" {
                        error-message "Configuration allowed in MST mode";
                        error-app-tag stp-invalid;
                    }
                    type uint8;
                    description
                        "MST hello time";
                }

                leaf max_age {
                    must "../../../STP/STP_LIST[keyleaf='GLOBAL']/mode='mst'" {
                        error-message "Configuration allowed in MST mode";
                        error-app-tag stp-invalid;
                    }
                    type uint8;
                    description
                        "MST max age";
                }

                leaf forward_delay {
                    must "../../../STP/STP_LIST[keyleaf='GLOBAL']/mode='mst'" {
                        error-message "Configuration allowed in MST mode";
                        error-app-tag stp-invalid;
                    }
                    type uint8;
                    description
                        "MST forward delay";
                }

                leaf hold_count {
                    must "../../../STP/STP_LIST[keyleaf='GLOBAL']/mode='mst'" {
                        error-message "Configuration allowed in MST mode";
                        error-app-tag stp-invalid;
                    }
                    type uint8;
                    description
                        "MST hold count";
                }
            }
        }

File: src/sonic-yang-models/yang-models/sonic-stp.yang [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+                when "../mode = 'mst'";
+                type uint8 {
+                    range "1..40";
+                }
+                default "20";
+                description "Maximum hops in MST region";
+            }
+        }
+
+        list vlan {
```
Comment (divyachandralekha) [inline src/sonic-yang-models/yang-models/sonic-stp.yang]: Vlan block should be like this 

 container STP_VLAN {
            list STP_VLAN_LIST {
                key "name";
                sonic-ext:custom-validation ValidateStpMaxInstance;
                leaf name {
                    type string;
                    description
                        "Vlan identifier";
                }

                leaf vlanid {
                    type uint16 {
                        range "1..4095" {
                            error-message "Vlan ID out of range";
                            error-app-tag vlanid-invalid;
                        }
                    }
                    description
                        "Vlan identifier number";
                }

                leaf enabled {
                    type boolean;
                    sonic-ext:custom-validation ValidateStpDisableVlan;
                    mandatory true;
                    description
                        "Spanning tree enabled/disabled on Vlan";
                }

                uses vlanModeAttr;
            }
        }

File: src/sonic-yang-models/yang-models/sonic-stp.yang [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+            leaf priority {
+                type uint16 {
+                    range "0..61440";
+                }
+                default "32768";
+                description "Bridge priority for this VLAN";
+            }
+        }
+
+        list interface {
```
Comment (divyachandralekha) [inline src/sonic-yang-models/yang-models/sonic-stp.yang]: This should be next container

File: src/sonic-yang-models/yang-models/sonic-stp.yang [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+                default "32768";
+                description "Bridge priority for this MST instance";
+            }
+            leaf vlan_list {
+                type string;
+                description "List of VLANs assigned to this MST instance";
+            }
+        }
+
+        list mst_port {
```
Comment (divyachandralekha) [inline src/sonic-yang-models/yang-models/sonic-stp.yang]: this should be next container

**Review by divyachandralekha — COMMENTED** [review]
Lets meet and discuss

File: src/sonic-yang-models/yang-models/sonic-stp.yang (hunk 5 above)
Comment (wajahatrazi) [reply src/sonic-yang-models/yang-models/sonic-stp.yang]: Please review, I have updated

File: src/sonic-yang-models/yang-models/sonic-stp.yang (hunk 4 above)
Comment (wajahatrazi) [reply src/sonic-yang-models/yang-models/sonic-stp.yang]: Please review, I have updated

File: src/sonic-yang-models/yang-models/sonic-stp.yang (hunk 3 above)
Comment (wajahatrazi) [reply src/sonic-yang-models/yang-models/sonic-stp.yang]: Please review, I have updated

File: src/sonic-yang-models/yang-models/sonic-stp.yang (hunk 2 above)
Comment (wajahatrazi) [reply src/sonic-yang-models/yang-models/sonic-stp.yang:177]: Please review, I have updated

File: src/sonic-yang-models/yang-models/sonic-stp.yang (hunk 1 above)
Comment (wajahatrazi) [reply src/sonic-yang-models/yang-models/sonic-stp.yang]: Please review, I have updated

File: src/sonic-yang-models/yang-models/sonic-stp.yang
Comment (wajahatrazi) [reply src/sonic-yang-models/yang-models/sonic-stp.yang:1]: Please review, I have updated

File: src/sonic-yang-models/yang-models/sonic-stp.yang (hunk 2 above)
Comment (divyachandralekha) [reply src/sonic-yang-models/yang-models/sonic-stp.yang:177]: module sonic-stp {
    namespace "http://github.com/sonic-net/sonic-stp";
    prefix stp;
    yang-version 1.1;

    import sonic-extension {
        prefix sonic-ext;
    }

    description "SONiC STP YANG model for PVST and MST configurations";

    revision 2024-11-24 {
        description "Version01 for combined PVST and MST configurations.";
    }

    grouping vlanModeAttr {
        leaf forward_delay {
            type uint8 {
                range "4..30" {
                    error-message "Invalid Forwarding Delay value.";
                }
            }
            units seconds;
            default 15;
            description
                "The delay used by STP bridges to transition root and
                designated ports to forwarding";
        }

        leaf hello_time {
            type uint8 {
                range "1..10" {
                    error-message "Invalid Hello Time value.";
                }
            }
            units seconds;
            default 2;
            description
                "The interval between periodic transmissions of
                configuration messages by designated ports";
        }

        leaf max_age {
            type uint8 {
                range "6..40" {
                    error-message "Invalid Maximum Age Time value.";
                }
            }
            units seconds;
            default 20;
            description
                "The maximum age of the information transmitted by the
                bridge when it is the root bridge";
        }

        leaf priority {
            type uint16 {
                range "0..61440" {
                    error-message "Invalid Bridge Priority value.";
                }
            }
            default 32768;
            description
                "The manageable component of the Bridge Identifier";
        }
    }

    grouping interfaceAttr {
        leaf path_cost {
            type uint64 {
                range "1..200000000" {
                    error-message "Invalid Port Path Cost value.";
                }
            }
            default 200;
            description
                "The port's contribution, when it is the Root Port,
                to the Root Path Cost for the Bridge";
        }

        leaf priority {
            type uint8 {
                range "0..240" {
                    error-message "Invalid Port Priority value.";
                }
            }
            default 128;
            description
                "The manageable component of the Port Identifier,
                also known as the Port Priority";
        }
    }


    container sonic-spanning-tree {

        container STP {
            list STP_LIST {
                max-elements 1;
                key "keyleaf";
                leaf keyleaf {
                    type enumeration {
                        enum GLOBAL;
                    }
                    description
                        "Key node identifier. It's value is always GLOBAL";
                }

                leaf mode {
                    type enumeration {
                        enum pvst;
                        enum mst;
                    }
                    mandatory true;
                    description
                        "Spanning tree mode";
                }

                leaf rootguard_timeout {
                    must "../../../STP/STP_LIST[keyleaf='GLOBAL']/mode!='mst'" {
                        error-message "Configuration not allowed in MST mode";
                        error-app-tag stp-invalid;
                    }

                    type uint16 {
                        range "5..600" {
                            error-message "Invalid Root-guard Timeout value.";
                        }
                    }
                    units seconds;
                    description
                        "Once superior BPDUs stop coming on the port, device
                        will wait for a period until root guard timeout before
                        moving the port to forwarding state";
                }

                uses vlanModeAttr;
            }
        }

        container STP_VLAN {
            list STP_VLAN_LIST {
                key "name";
                leaf name {
                    type string;
                    description
                        "Vlan identifier";
                }

                leaf vlanid {
                    type uint16 {
                        range "1..4095" {
                            error-message "Vlan ID out of range";
                            error-app-tag vlanid-invalid;
                        }
                    }
                    description
                        "Vlan identifier number";
                }

                leaf enabled {
                    type boolean;
                    mandatory true;
                    description
                        "Spanning tree enabled/disabled on Vlan";
                }

                uses vlanModeAttr;
            }
        }

        container STP_VLAN_PORT {
            list STP_VLAN_PORT_LIST {
                key "vlan-name ifname";

                leaf vlan-name {
                    type leafref {
                        path "../../../STP_VLAN/STP_VLAN_LIST/name";
                    }
                    description
                        "Reference to Vlan";
                }

                leaf ifname {
                    type leafref {
                        path "../../../STP_PORT/STP_PORT_LIST/ifname";
                    }
                    description
                        "Reference to Ethernet interface or PortChannel";
                }

                uses interfaceAttr;
            }
        }

        container STP_PORT {
            list STP_PORT_LIST {
                key "ifname";
                sonic-ext:dependent-on "STP_LIST";
                leaf ifname {
                    type string;
                    description
                        "Reference to Ethernet interface or PortChannel";
                }

                leaf enabled {
                    type boolean;
                    mandatory true;
                    description
                        "Spanning tree enabled/disabled on Interface";
                }

                leaf root_guard {
                    type boolean;
                    description
                        "Enable/Disable Root guard on port";
                }

                leaf bpdu_guard {
                    type boolean;
                    description
                        "Enable/Disable port BPDU guard";
                }

                leaf bpdu_guard_do_disable {
                    type boolean;
                    description
                        "Port to be disabled when it receives a BPDU";
                }

                leaf uplink_fast {
                    type boolean;
                    description
                        "Enable/Disable uplink-fast on port";
                }

                leaf portfast {
                    must "current()!='true' or ../../../STP/STP_LIST[keyleaf='GLOBAL']/mode='pvst'" {
                        error-message "Configuration not allowed in MST mode";
                        error-app-tag stp-invalid;
                    }
                    type boolean;
                    description
                        "Enable/Disable portfast on port";
                }

                uses interfaceAttr;

                // For MST
                leaf edge_port {
                    type boolean;
                    description
                        "Enable/Disable Edge-port on interface";
                }

                leaf link_type {
                    type enumeration {
                        enum auto;
                        enum shared;
                        enum point-to-point;
                    }
                    description
                        "Specifies the interface's link type. Permissible values
                        are 'shared', 'point-to-point' and 'auto'";
                }
            }
        }

        container STP_MST {
            list STP_MST_LIST {
                max-elements 1;
                key "keyleaf";
                sonic-ext:dependent-on "STP_LIST";

                leaf keyleaf {
                    type enumeration {
                        enum GLOBAL;
                    }
                    description
                        "Key node identifier. It's value is always GLOBAL";
                }

                leaf name {
                    must "../../../STP/STP_LIST[keyleaf='GLOBAL']/mode='mst'" {
                        error-message "Configuration allowed in MST mode";
                        error-app-tag stp-invalid;
                    }
                    type string;
                    description
                        "MST Region name";
                }

                leaf revision {
                    must "../../../STP/STP_LIST[keyleaf='GLOBAL']/mode='mst'" {
                        error-message "Configuration allowed in MST mode";
                        error-app-tag stp-invalid;
                    }
                    type uint32;
                    description
                        "MST Revision number";
                }

                leaf max_hops {
                    must "../../../STP/STP_LIST[keyleaf='GLOBAL']/mode='mst'" {
                        error-message "Configuration allowed in MST mode";
                        error-app-tag stp-invalid;
                    }
                    type uint8;
                    description
                        "MST Max hops";
                }

                leaf hello_time {
                    must "../../../STP/STP_LIST[keyleaf='GLOBAL']/mode='mst'" {
                        error-message "Configuration allowed in MST mode";
                        error-app-tag stp-invalid;
                    }
                    type uint8;
                    description
                        "MST hello time";
                }

                leaf max_age {
                    must "../../../STP/STP_LIST[keyleaf='GLOBAL']/mode='mst'" {
                        error-message "Configuration allowed in MST mode";
                        error-app-tag stp-invalid;
                    }
                    type uint8;
                    description
                        "MST max age";
                }

                leaf forward_delay {
                    must "../../../STP/STP_LIST[keyleaf='GLOBAL']/mode='mst'" {
                        error-message "Configuration allowed in MST mode";
                        error-app-tag stp-invalid;
                    }
                    type uint8;
                    description
                        "MST forward delay";
                }

                leaf hold_count {
                    must "../../../STP/STP_LIST[keyleaf='GLOBAL']/mode='mst'" {
                        error-message "Configuration allowed in MST mode";
                        error-app-tag stp-invalid;
                    }
                    type uint8;
                    description
                        "MST hold count";
                }
            }
        }
        container STP_MST_INST {
            list STP_MST_INST_LIST {
                key "instance";

                leaf instance {
                    type uint16;
                    description
                        "Instance identifier";
                }

                leaf-list vlan {
                    type string;
                    description
                        "Vlan list";
                }

                leaf bridge_priority {
                    type uint16 {
                        range "0..61440" {
                            error-message "Invalid Bridge Priority value.";
                        }
                    }
                    description
                        "The manageable component of the Bridge Identifier";
                }
            }
        }

        container STP_MST_PORT {
            list STP_MST_PORT_LIST {
                key "inst_id ifname";

                leaf inst_id {
                    type leafref {
                        path "../../../STP_MST_INST/STP_MST_INST_LIST/instance";
                    }
                    description
                        "Reference to MST Instance";
                }

                leaf ifname {
                    type leafref {
                        path "../../../STP_PORT/STP_PORT_LIST/ifname";
                    }

                    description
                        "Reference to Ethernet interface or PortChannel";
                }
                uses interfaceAttr;
            }
        }
    }  
}

**Pull request closed without merging**